src/Checkpointer.jl

```diff
@@ -468,4 +509,187 @@ function restore!(
+    ignore::Set{Symbol} = Set{Symbol}(),
+)
+    stack = []
+    push!(stack, obj)
+    return FieldIterator{typeof(is_leaf), typeof(stop)}(stack, is_leaf, stop, ignore)
+end
+
+"""
+    iterate(itr::FieldIterator)
+
+Advance the iterator to obtain the next field in the field hierarchy of the
+object. If no fields remain, `nothing` is returned. Otherwise, a 2-tuple of the
+next element and the new iteration state is returned.
+"""
+function Base.iterate(itr::FieldIterator)
+    stack = itr.stack
+    ignore = itr.ignore
+    while !isempty(stack)
+        obj = pop!(stack)
+        itr.stop(obj) && continue
+        itr.is_leaf(obj) && return (obj, nothing)
+        T = typeof(obj)
+        fields = filter(x -> !(x in ignore), fieldnames(T))
+        for p in fields
+            push!(stack, getfield(obj, p))
+        end
+    end
+    return nothing
+end
+
+# Functions needed to implement the iterator interface
+# see https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-iteration
+Base.iterate(iter::FieldIterator, _) = iterate(iter)
+Base.IteratorEltype(::FieldIterator) = Base.EltypeUnknown()
+Base.IteratorSize(::FieldIterator) = Base.SizeUnknown()
+Base.isdone(itr::FieldIterator, state...) = isempty(itr.stack)
+
+function stop(
+    ::Union{
+        AbstractTimeVaryingInput,
+        ClimaComms.AbstractCommsContext,
+        ClimaComms.AbstractDevice,
+        UnionAll,
+        DataType,
+    },
+)
+    return true # Stop recursing here
+end
+
+# Catch all for everything else
+stop(obj) = false
+
+# Needed for saving data
+is_leaf(::Union{CC.DataLayouts.AbstractData, AbstractArray}) = true
+
+# Needed for error handling
+is_leaf(::Union{StaticArrays.StaticArray, Number, UnitRange, LinRange, Symbol}) = true
+
+# Needed for error handling
+is_leaf(::Dict) = true
+
+# Needed for error handling
+is_leaf(::Union{Dates.DateTime, Dates.UTInstant, Dates.Millisecond}) = true
+
+# Catch all for everything else
+is_leaf(::Any) = false
+
+"""
+    create_atmos_cache_itr(cache)
+
+Create an iterator over the atmos cache of objects that should be saved.
+"""
+function create_atmos_cache_itr(cache)
+    return FieldIterator(
+        cache,
+        is_leaf;
+        stop = stop,
+        ignore = Set([
+            :rc,
+            :params,
+            :ghost_buffer,
+            :hyperdiffusion_ghost_buffer,
+            :data_handler,
+            :graph_context,
+            :dt,
+        ]),
+    )
+end
+
+"""
+    make_atmos_cache(cache)
+
+Collect the data in the `cache` that is needed for restarting the atmos cache.
+
+Objects that share the same object ID are not copied, but a reference to the
+object is used instead.
+"""
+function make_atmos_cache(cache)
+    atmos_cache_itr = create_atmos_cache_itr(cache)
+    cache_vec = [] # Elements of the vector can be any type
+    obj_id_to_idx = Dict{UInt64, Int}()
+    for (i, obj) in enumerate(atmos_cache_itr)
+        obj_id = objectid(obj)
+        if obj_id in keys(obj_id_to_idx)
```

Comment:
I don't understand this part - does `obj_id_to_idx` check if we've already added this object to the cache vec? It would be good to have some comments to clarify